### PR TITLE
Darwin: Only set FD_CLOEXEC when creating a socket FD

### PR DESCRIPTION
**Pull request by carlhoerberg**
Not when an existing FD is passed to Socket.new

### CODE REVIEW

**Comment by carlhoerberg** [comment]
Follow up on https://github.com/crystal-lang/crystal/pull/14632#pullrequestreview-2080836656

**Review by straight-shoota — APPROVED** [review]

File: src/crystal/system/unix/socket.cr [hunk 1]
```diff
@@ -15,15 +15,15 @@ module Crystal::System::Socket
...
     raise ::Socket::Error.from_errno("Failed to create socket") if fd == -1
-    fd
-  end
-
-  private def initialize_handle(fd)
     {% unless LibC.has_constant?(:SOCK_CLOEXEC) %}
       # Forces opened sockets to be closed on `exec(2)`. Only for platforms that don't
       # support `SOCK_CLOEXEC` (e.g., Darwin).
-      LibC.fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
+      fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
```
Comment (straight-shoota) [inline src/crystal/system/unix/socket.cr:21]: **issue:** When using the instance method, there's no `fd` argument:

```suggestion
      fcntl(LibC::F_SETFD, LibC::FD_CLOEXEC)
```

**Review by straight-shoota — CHANGES_REQUESTED** [review]

**Comment by ysbaddaden** [comment]
I notice some issues with this change: when we create the socket the `@handle` ivar isn't set and calling `#fcntl` will fail (see CI) but we also set CLOEXEC for plain `fd` (e.g. `Socket.new(fd, ...)` which is used in `UNIXSocket.pair` for example.

**Comment by straight-shoota** [comment]
> when we create the socket the `@handle` ivar isn't set and calling `#fcntl` will fail (see CI)

Yeah I think we need to use `LibC.fcntl(fd, ...)` like before to fix the current failure. The file descriptor is not assigned yet, so we cannot use `#fcntl`.

> we also set CLOEXEC for plain `fd` (e.g. `Socket.new(fd, ...)` as called from `UNIXSocket.pair` or `Socket#accept?` for example.

Yes, that's the current behaviour. But I believe it's wrong. We probably shouldn't autoclose a file descriptor that's coming from outside the process. Only close what we opened ourselves.
That means all uses in stdlib like `Socket#accept?` need to be updated to explicitly set close on exec.

However, I'm wondering if this isn't too bad a silent breaking change in behaviour... 🤔

**Comment by ysbaddaden** [comment]
I agree: it may not be the best API, but security-wise it doesn't sound like a bad practice to always close file descriptors on exec... and maybe always calling `fcntl` ain't a bad idea instead of setting `O_CLOEXEC`. If you need it to not be closed on exec, you can always call `fcntl` to disable the flag.

File: src/crystal/system/unix/socket.cr [hunk 2]
```diff
@@ -15,15 +15,16 @@ module Crystal::System::Socket
...
-    fd
-  end
-
-  private def initialize_handle(fd)
     {% unless LibC.has_constant?(:SOCK_CLOEXEC) %}
       # Forces opened sockets to be closed on `exec(2)`. Only for platforms that don't
       # support `SOCK_CLOEXEC` (e.g., Darwin).
-      LibC.fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
+      r = LibC.fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
+      raise ::Socket::Error.from_errno("fcntl() failed") if r == -1
```
Comment (Sija) [inline src/crystal/system/unix/socket.cr:22]: Suggestion: avoid using one letter names for clarity.

```suggestion
      ret = LibC.fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
      raise ::Socket::Error.from_errno("fcntl() failed") if ret == -1
```

**Review by straight-shoota — APPROVED** [review]

**Comment by ysbaddaden** [comment]
The following methods (at least) must also be fixed to set `FD_CLOEXEC` for darwin targets:

- `Socket#accept?`
- `UNIXSocket#accept?`
- `UNIXSocket.pair`

Why Apple wouldn't implement the `O_CLOEXEC` flag is beyond me :face_exhaling:

**Review by ysbaddaden — CHANGES_REQUESTED** [review]
The following methods (at least) must also be fixed to set `FD_CLOEXEC` for darwin targets:

- `Socket#accept?`
- `UNIXSocket#accept?`
- `UNIXSocket.pair`

Maybe there should be a system method to set `FD_CLOEXEC` on a given `fd` to avoid repeating the same `LibC.fcntl` call? Something like `Crystal::System::FileDescriptor.set_close_on_exec(fd : Int32) : Nil`.

Why Apple wouldn't implement the ~~`O_CLOEXEC`~~ `SOCK_CLOEXEC` flag is beyond me :face_exhaling:

**Comment by straight-shoota** [comment]
Yeah, I suppose we should also setup specs for those methods.

@carlhoerberg If you want to go on, please continue. But we'll be happy to take over these final steps if you prefer.

**Comment by carlhoerberg** [comment]
Go for it! :)

**Pull request closed without merging**